src/main/java/org/cryptomator/cloudaccess/vaultformat8/VaultFormat8ProviderDecorator.java

```diff
@@ -98,26 +103,74 @@ public VaultFormat8ProviderDecorator(CloudProvider delegate, CloudPath dataDir,
 
 	@Override
 	public CompletionStage<CloudItemMetadata> write(CloudPath file, boolean replace, InputStream data, ProgressListener progressListener) {
-		return CompletableFuture.failedFuture(new UnsupportedOperationException("not implemented"));
+		return getC9rPath(file).thenCompose(filePath -> {
+			try (var src = Channels.newChannel(data);
+				 var encryptingChannel = new EncryptingReadableByteChannel(src, cryptor);
+				 var encryptedIn= Channels.newInputStream(encryptingChannel)) {
+				return delegate.write(filePath, replace, encryptedIn, progressListener);
+			} catch (IOException e) {
+				return CompletableFuture.failedFuture(e);
+			}
+		});
 	}
 
 	@Override
 	public CompletionStage<CloudPath> createFolder(CloudPath folder) {
-		return CompletableFuture.failedFuture(new UnsupportedOperationException("not implemented"));
+		BiFunction<CloudPath, Throwable, CompletableFuture<CloudPath>> handler = (result, exception) -> {
+			if (exception == null) {
+				return CompletableFuture.completedFuture(result);
+			} else if (exception instanceof FileAlreadyExistsException) {
+				return CompletableFuture.completedFuture(result);
```

Comment:
I don't think there is a `result` in case of an `exception`. Use `folder` instead?

Reply:
Is this covered by tests?